### ISSUE TITLE
Event triggered for else-if-case that should not be reached

### ISSUE THREAD

**Issue by AnHeuermann** [OpenModelica/OpenModelica#7657]
### Description

When having an if equation with `if` and `else if` for all relations state events are triggered, even if non should be generated because we are still in the first case.
Not sure if there is already a dedicated ticket for this issue on track.

### Steps to Reproduce
```Modelica
model mwe
  Real v;
  Real aux;
equation
  aux=sin(20*time);
  v = if time < 0.1 then 0 else if aux < 0.5 then aux else 0.5;
end mwe;
```

When simulating with `LOG_EVENTS` we see that there is an state event detected at `time=0.026`, because relation `aux<0.5` changes. But at that time the first case of the if equation is still valid, so OpenModelcia shouldn't even check the second case.

It is of course fine to update the relation but an event shouldn't be triggered.

```
messages = "LOG_EVENTS        | info    | status of relations at time=0
|                 | |       | | [1] (pre:  true)  true = time < 0.1
|                 | |       | | [2] (pre:  true)  true = aux < 0.5
LOG_EVENTS        | info    | status of zero crossings at time=0
|                 | |       | | [1] (pre:  0)  1 = time < 0.1
|                 | |       | | [2] (pre:  0)  1 = aux < 0.5
LOG_SUCCESS       | info    | The initialization finished successfully without homotopy method.
LOG_EVENTS        | info    | state event at time=0.0261799387895
|                 | |       | | [2] aux < 0.5
LOG_EVENTS        | info    | state event at time=0.100000000111
|                 | |       | | [1] time < 0.1
LOG_EVENTS        | info    | state event at time=0.130899693909
|                 | |       | | [2] aux < 0.5
LOG_EVENTS        | info    | state event at time=0.340339204148
|                 | |       | | [2] aux < 0.5
```

### Version and OS

 - OpenModelica Version: v1.18.0-dev-232-g187041202b (64-bit)
 - OS: Windows 10

**Comment by AnHeuermann** [comment]
One way to fix this could be to change the zero crossing function a bit. For relations in `else if` we need to check if the previous case is `false` before updating the zero crossing. I'm not sure if we store this information inside the backend / simcode. Adapting the code generation would be easy.

For this example we would need to change
```C
int mwe_function_ZeroCrossings(DATA *data, threadData_t *threadData, double *gout)
{
  [...]
  tmp0 = LessZC(data->localData[0]->timeValue, 0.1, data->simulationInfo->storedRelations[0]);
  gout[0] = (tmp0) ? 1 : -1;
  tmp1 = LessZC(data->localData[0]->realVars[0] /* aux variable */, 0.5, data->simulationInfo->storedRelations[1]);
  gout[1] = (tmp1) ? 1 : -1;
  [...]
}
```
to
```C
int mwe_function_ZeroCrossings(DATA *data, threadData_t *threadData, double *gout)
{
  [...]
  tmp0 = LessZC(data->localData[0]->timeValue, 0.1, data->simulationInfo->storedRelations[0]);
  gout[0] = (tmp0) ? 1 : -1;
  tmp1 = LessZC(data->localData[0]->realVars[0] /* aux variable */, 0.5, data->simulationInfo->storedRelations[1]);
  gout[1] = (!gout[0] && tmp1) ? 1 : -1;
  [...]
}
```

**Comment by AnHeuermann** [comment]
Btw. using

```Modelica
model mwe
  Real v;
  Real aux;
equation
  aux=sin(20*time);
  if time < 0.1 then
    v = 0;
  elseif aux < 0.5 then
    v = aux;
  else
    v = 0.5;
  end if;
end mwe;
```

has the same wrong behavior.

**Comment by casella** [comment]
According to my understanding, zero-crossing functions shouldn't be discontinuous, just in case the solver uses something more sophisticated than bisection to locate the event time. For a relation such as `x > y` we should return `x - y`, not `if x > y then 1 else -1`.

See also [#6411](https://trac.openmodelica.org/OpenModelica/ticket/6411#comment:9)

**Comment by AnHeuermann** [comment]
I'm not sure if I understand your comment.
The point here is not that the zero crossing function itself is continuous or not, but that an event is triggered where it shouldn't, because the second zero crossing doesn't know that it is from an `elseif` and needs to "wait" for the relation from the previous `if`.

**Comment by casella** [comment]
If I understand it correctly, the code fragments you posted generate step ZC functions. In that case, it is relatively straightforward to implement what you propose. I'm not sure it still is if you actually want to generated proper continuous ZC functions.

**Comment by casella** [comment]
As I understand it, we should always compute continuous ZC functions for all relations in the model. Then, we should only "show" the solver the ones that are active. I'm not sure what is the best way to mask out the ones that the solver shouldn't bother about, maybe they should simply return the previous result if they are not actually involved in event generation.

Maybe you can invent a new Boolean algebra for continuous ZC functions, so that any complex conditions can be handled by reducing it to an equivalent _continuous_ ZC function by suitably combining the individual ZC functions of the constituent relations. I'm not sure whether this is theoretically possible, you may give it a thought.

**Comment by AnHeuermann** [comment]
Yes, continuous zero crossing should be a nice thing, but for now we only have discontinuous ones available.  This is definitely something we need to keep in mind if we want to change that to some continuous way some day.
But for this ticket I would stick to our usual discontinuous (`+1` or `-1`) approach on handling zero crossings.

And yes, it is straight forward to implement in code generation, if the backend knows that this zero-crossing is from an `elseif`.

**Comment by casella** [comment]
OK. I would just avoid to spend too much time optimizing step-ZCFs, which are already sub-optimal by themselves.

**Comment by casella** [comment]
@AnHeuermann, could you finalize this issue based on your last comment?

**Comment by AnHeuermann** [comment]
Not really. We would need to get an ordering an dependencies into the zero-crossing equations.
I don't know how we should do that in the backend. 
Maybe @kabdelhak can give a rough estimation on how much work it would be to add this to the backend.

But I'm not sure if this ticket is that important. Al that is happening is that we waste some time in an event iteration. The result itself is still correct.
